server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java

```diff
@@ -21,38 +21,195 @@
+            MessageId messageId = sendMdn(creationMDNEntry, mailboxSession);
+            return SetMessagesResponse.builder()
+                .mdnSent(creationMDNEntry.getCreationId(), messageId);
+        } catch (MessageNotFoundException e) {
+            return SetMessagesResponse.builder()
+                .MDNNotSent(creationMDNEntry.getCreationId(),
+                    SetError.builder()
+                        .description(String.format("Message with id %s not found. Thus could not send MDN.",
+                            creationMDNEntry.getValue().getMessageId().serialize()))
+                        .type("invalidArgument")
+                        .build());
+
+        } catch (Exception e) {
+            LOGGER.error("Error while sending MDN", e);
+            return SetMessagesResponse.builder()
+                .MDNNotSent(creationMDNEntry.getCreationId(),
+                    SetError.builder()
+                        .description(String.format("Could not send MDN %s", creationMDNEntry.getCreationId().getId()))
+                        .type("error")
+                        .build());
+        }
     }
 
-    public SetMessagesResponse handleMDNCreation(SetMessagesRequest request) {
-        SetMessagesResponse.Builder builder = SetMessagesResponse.builder();
+    public MessageId sendMdn(ValueWithId.CreationMDNEntry creationMDNEntry, MailboxSession mailboxSession) throws MailboxException, IOException, MessagingException, ParseException, MessageNotFoundException {
+        JmapMDN mdn = creationMDNEntry.getValue();
+        Message originalMessage = retrieveOriginalMessage(mdn, mailboxSession);
+        MDNReport mdnReport = generateReport(mdn, originalMessage, mailboxSession);
+        List<MessageAttachment> reportAsAttachment = ImmutableList.of(convertReportToAttachment(mdnReport));
+        User user = User.fromUsername(mailboxSession.getUser().getUserName());
+
+        Message mdnAnswer = generateMDNMessage(originalMessage, mdn, mdnReport, user);
+
+        Flags seen = new Flags(Flags.Flag.SEEN);
+        MessageFactory.MetaDataWithContent metaDataWithContent = messageAppender.appendMessageInMailbox(mdnAnswer,
+            getOutbox(mailboxSession), reportAsAttachment, seen, mailboxSession);
 
-        request.getSendMDN()
-            .forEach(creationMDNEntry -> builder.MDNNotSent(creationMDNEntry.getCreationId(),
-                SetError.builder()
-                    .description(String.format("Could not send MDN %s", creationMDNEntry.getCreationId().getId()))
-                    .type("Not implemented yet")
-                    .build()));
+        messageSender.sendMessage(metaDataWithContent,
+            Envelope.fromMime4JMessage(mdnAnswer), mailboxSession);
 
-        return builder.build();
+        return metaDataWithContent.getMessageId();
     }
+
+    public Message generateMDNMessage(Message originalMessage, JmapMDN mdn, MDNReport mdnReport, User user) throws ParseException, IOException {
```

Comment:
Maybe private, no? (same for other methods)

Reply:
Looks like a bad IDE setting...